File: src/PYTHON/gRPC_TypedClient.py
import sys
import os
import grpc

# Đảm bảo Python nhận diện đúng thư mục để import các file sinh ra từ proto
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
import judge_pb2
import judge_pb2_grpc

def run():
    # 1. Kết nối đến gRPC Server (plaintext, không TLS) tại cổng 2240
    server_address = "36.50.135.242:2240"
    channel = grpc.insecure_channel(server_address)
    
    # Chú ý sử dụng TypedJudgeServiceStub thay vì JudgeServiceStub cũ
    stub = judge_pb2_grpc.TypedJudgeServiceStub(channel)

    student_code = "B22DCCN292"
    question_alias = "xw74a56O"

    # 2. Gửi yêu cầu RequestTyped để nhận đề bài
    print("Đang gửi yêu cầu RequestTyped...")
    request = judge_pb2.TypedJudgeRequest(
        student_code=student_code,
        question_alias=question_alias
    )
    
    try:
        response = stub.RequestTyped(request)
        print("Đã nhận phản hồi từ Server:")
        print(f"Request ID: {response.request_id}")
        
        # 3. Phân tích dữ liệu text_batch
        if not response.HasField("text_batch"):
            print("Lỗi: Server không trả về kiểu dữ liệu text_batch!")
            return
            
        text_batch = response.text_batch
        print(f"Mode nhận được: {text_batch.mode}")
        
        # Các nhãn cần đếm
        targets = ["account", "payment", "refund", "shipping"]
        
        # Tạo Dictionary để lưu số lượng đếm được cho mỗi nhãn
        counts = {t: 0 for t in targets}
        
        # Duyệt qua từng entry (mỗi phần tử trong list entries)
        for entry in text_batch.entries:
            # Chuyển thành chữ thường để so sánh không phân biệt hoa/thường
            entry_lower = entry.lower()
            for t in targets:
                # Kiểm tra nếu nhãn xuất hiện trong chuỗi entry
                if t in entry_lower:
                    counts[t] += 1
                    
        # Lọc ra các nhãn có count > 0 và sắp xếp theo thứ tự ABC
        values = []
        final_counts = {}
        
        # Hàm sorted(targets) đảm bảo danh sách đã được sắp xếp tăng dần theo ABC
        for t in sorted(targets):
            if counts[t] > 0:
                values.append(t)
                final_counts[t] = counts[t]
                
        print(f"Danh sách nhãn có xuất hiện (values): {values}")
        print(f"Số lượng tương ứng (counts): {final_counts}")

        # 4. Đóng gói vào đối tượng TextBatchAnswer
        text_answer = judge_pb2.TextBatchAnswer(
            values=values,
            counts=final_counts
        )
        
        # Tạo request submit tổng thể
        submit_req = judge_pb2.TypedSubmitRequest(
            student_code=student_code,
            question_alias=question_alias,
            request_id=response.request_id,
            text_batch_answer=text_answer
        )
        
        print("\nĐang gửi kết quả SubmitTyped...")
        submit_res = stub.SubmitTyped(submit_req)
        print("Kết quả từ Server:")
        print(f"Status: {submit_res.status}")
        print(f"Message: {submit_res.message}")
        
    except grpc.RpcError as e:
        print(f"Lỗi gRPC xảy ra: {e.code()} - {e.details()}")

if __name__ == "__main__":
    run()

File: src/main/java/com/mycompany/soap/gRPC_TypedClient.java
package com.mycompany.soap;

import GRPC.TypedJudgeServiceGrpc;
import GRPC.TypedJudgeRequest;
import GRPC.TypedJudgeResponse;
import GRPC.TypedSubmitRequest;
import GRPC.TypedSubmitResponse;
import GRPC.TextBatchAnswer;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class gRPC_TypedClient {
    public static void main(String[] args) {
        String studentCode = "B22DCCN292";
        String questionAlias = "xw74a56O";

        // 1. Khởi tạo Channel kết nối tới gRPC Server (plaintext, không TLS)
        ManagedChannel channel = ManagedChannelBuilder.forAddress("36.50.135.242", 2240)
                .usePlaintext()
                .build();

        // Khởi tạo Stub (Blocking)
        TypedJudgeServiceGrpc.TypedJudgeServiceBlockingStub stub = TypedJudgeServiceGrpc.newBlockingStub(channel);

        // 2. Gửi RequestTyped
        TypedJudgeRequest request = TypedJudgeRequest.newBuilder()
                .setStudentCode(studentCode)
                .setQuestionAlias(questionAlias)
                .build();
                
        System.out.println("Đang gửi yêu cầu RequestTyped...");
        TypedJudgeResponse response = stub.requestTyped(request);
        
        System.out.println("Đã nhận phản hồi từ Server, Request ID: " + response.getRequestId());
        
        // Kiểm tra xem server có trả về text_batch không
        if (!response.hasTextBatch()) {
            System.out.println("Lỗi: Không nhận được text_batch!");
            channel.shutdown();
            return;
        }

        System.out.println("Mode nhận được: " + response.getTextBatch().getMode());

        // 3. Phân tích văn bản
        List<String> entries = response.getTextBatch().getEntriesList();
        String[] targets = {"account", "payment", "refund", "shipping"};
        Map<String, Integer> counts = new HashMap<>();
        
        // Khởi tạo số đếm bằng 0
        for (String t : targets) {
            counts.put(t, 0);
        }

        // Đếm số lần xuất hiện không phân biệt hoa thường
        for (String entry : entries) {
            String lowerEntry = entry.toLowerCase();
            for (String t : targets) {
                if (lowerEntry.contains(t)) {
                    counts.put(t, counts.get(t) + 1);
                }
            }
        }
        
        // 4. Lọc bỏ các nhãn có số lượng = 0 và sắp xếp
        List<String> values = new ArrayList<>();
        Map<String, Integer> finalCounts = new HashMap<>();
        
        // Mảng targets đã định nghĩa là "account", "payment", "refund", "shipping"
        // đã theo chuẩn thứ tự ABC nên ta chỉ việc thêm vào. Để an toàn, gọi Arrays.sort
        Arrays.sort(targets);
        
        for (String t : targets) {
            if (counts.get(t) > 0) {
                values.add(t);
                finalCounts.put(t, counts.get(t));
            }
        }
        
        System.out.println("Values: " + values);
        System.out.println("Counts: " + finalCounts);
        
        // Đóng gói vào đối tượng TextBatchAnswer
        TextBatchAnswer answer = TextBatchAnswer.newBuilder()
                .addAllValues(values)
                .putAllCounts(finalCounts)
                .build();
                
        // 5. Đóng gói vào SubmitRequest
        TypedSubmitRequest submitRequest = TypedSubmitRequest.newBuilder()
                .setStudentCode(studentCode)
                .setQuestionAlias(questionAlias)
                .setRequestId(response.getRequestId())
                .setTextBatchAnswer(answer)
                .build();
                
        System.out.println("\nĐang gửi kết quả SubmitTyped...");
        TypedSubmitResponse submitResponse = stub.submitTyped(submitRequest);
        System.out.println("Kết quả từ Server:");
        System.out.println("Status: " + submitResponse.getStatus());
        System.out.println("Message: " + submitResponse.getMessage());

        // Đóng channel sau khi hoàn thành
        channel.shutdown();
    }
}
